idseq_pipeline/commands/host_filtering_functions.py

```diff
@@ -240,30 +240,83 @@ def uncompressed(s3genome):
+            # Hmm. I understand the desire to create a single function for fetching any reference we might need,
+            # but the functionality of these two functions is actually quite different.  One of them handles
+            # resiliently situations like the destination existing/not existing, being a folder/not a folder, etc.
+            # The other expands an archive into an existing folder, and needs to thoroughly cleanup in case
+            # something does not go as planned, because of the prefetching pattern of use.  Also, this function
+            # will fetch whichever of the "tar" or "tar.gz" versions of the archive actually exist in s3,
+            # even if the other version has been specified.  So the two functions are resilient in different ways.
             install_s3mi()
             tarfile = uncompressed(s3genome)
             try:
                 execute_command("s3mi cat {tarfile} | tar xvf - -C {refdir}".format(tarfile=tarfile, refdir=REF_DIR))
+                assert os.path.isdir(genome_dir)
             except:
                 if tarfile != s3genome:
                     # The uncompressed version doesn't exist.   This is much slower, but no choice.
+                    execute_command("rm -rf {}".format(genome_dir))
                     execute_command("s3mi cat {s3genome} | tar xvfz - -C {refdir}".format(s3genome=s3genome, refdir=REF_DIR))
+                    assert os.path.isdir(genome_dir)
                 else:
+                    # Okay, may be s3mi is broken.  We'll try aws cp next.
                     raise
         except:
-            execute_command("aws s3 cp --quiet {s3genome} - | tar xvf - -C {refdir}".format(s3genome=s3genome, refdir=REF_DIR))
-        write_to_log("downloaded index")
-    assert os.path.isdir(genome_dir)
+            try:
+                execute_command("rm -rf {}".format(genome_dir))
+                execute_command("aws s3 cp --quiet {s3genome} - | tar xvf - -C {refdir}".format(s3genome=s3genome, refdir=REF_DIR))
+                assert os.path.isdir(genome_dir)
+            except:
+                write_to_log("Failed to download index {}, it might not exist.".format(s3genome))
+                genome_dir = None
+                # Note we do not reraise the exception here, just print it.
+                traceback.print_exc()
+        if genome_dir:
+            write_to_log("downloaded index {}".format(s3genome))
     return genome_dir
 
 
 def fetch_genome(s3genome, mutex=threading.RLock(), mutexes={}): #pylint: disable=dangerous-default-value
+    """Fetch and expand genome archive from s3 into local dir.  Return that local dir.
+    If all retries fail, return None.  If already downloaded, return right away.
+    If a fetch of the same genome is already in progress on another thread, wait for it.
+
+    Typical use:
+
+            fruitfly_dir = fetch_genome("s3://fruitfly_genome.tar")
+            assert fruitfly_dir != None
```

Comment:
We can get rid of this assert business as soon as we address issue https://github.com/chanzuckerberg/idseq-pipeline/issues/173